Improve if/return optimisation without a trailing return after if block (with return statement)

avdg opened the pull request:
Only apply the optimization in question when a return keyword can be removed.

\<old intro>
Still need to consider if this optimization is totally useless.

Pro case:
- Joins multiple statements into 1 statement, which could be useful for `return cond ? if_body : (alternative_or_tail)`

Contra case:
- Adding undefined to the last statement makes it produce longer code.
- Current transformer typically creates longer code with given code cases
- Other transformers can join multiple statements into 1 as well, I'm not sure if there are cases where it isn't

</old intro>

avdg commented on test/compress/if_return.js:
```diff
@@ -171,7 +171,7 @@ if_return_7: {
     }
     expect: {
```
Any way to write this shorter? To my knowledge I can't write this shorter.

avdg commented:
@kzc I guess I made you happy now :-) https://github.com/mishoo/UglifyJS2/issues/866#issuecomment-157040203

avdg commented:
This patch seems to squeeze 11 chars off from jquery 3.1.1

kzc commented:
Contra case I found in the wild added 500 bytes. Adapted below:

$ cat if_test.js 
```
function f(e) {
    if (2 == e) return foo();
    if (3 == e) return bar();
    if (4 == e) return baz();
    fail(e);
}
```
**Without PR**

$ uglifyjs if_test.js -cm
```
function f(f){return 2==f?foo():3==f?bar():4==f?baz():void fail(f)}
```
**With PR**

$ bin/uglifyjs if_test.js -cm
```
function f(f){if(2==f)return foo();if(3==f)return bar();if(4==f)return baz();fail(f)}
```

avdg commented:
nice one

avdg commented:
So obviously there is something to improve on.

avdg commented:
Actually I would use the example below because your example can be sneakingly be improved with putting things in an array.

```js
function f(e) {
    if (a(e)) return foo();
    if (b(e)) return bar();
    if (c(e)) return baz();
    fail(e);
}
```

kzc commented:
> your example can be sneakingly be improved with putting things in an array

same difference - uglify will make no such improvement

mishoo commented:
IIRC, @kzc's example was exactly the reason for the complication.  It would be nice if we could still handle that nicely.

I'm not against this patch though.  Everything which removes code is probably good. :-)  Leaving it up to you guys.

kzc commented:
With this patch every additional if+return statement adds 10 bytes if there is no trailing return at end of block. Ideally we should detect such cases.

Testing on local JS files I found that this PR is effective 90% of the time, but when it isn't it erases all the gains of the other files and then some.

avdg commented:
The optimization only works on cases with at least 2 return statements, where 1 return statement gets removed from the code.

So I guess we need to detect that if we want the shortest code.

kzc commented:
At least two if+return statements with no trailing return in the block.

avdg commented:
```diff
diff --git a/lib/compress.js b/lib/compress.js
index 177fccb..7e4e967 100644
--- a/lib/compress.js
+++ b/lib/compress.js
@@ -541,7 +541,11 @@ merge(Compressor.prototype, {
             var multiple_if_returns = has_multiple_if_returns(statements);
             var in_lambda = self instanceof AST_Lambda;
             var ret = []; // Optimized statements, build from tail to front
+            console.log("=====");
+            console.log(statements.map(a=>a.TYPE));
+            console.log("-----");
             loop: for (var i = statements.length; --i >= 0;) {
+                console.log(ret.map(a=>a.TYPE));
                 var stat = statements[i];
                 switch (true) {
                   case (in_lambda && stat instanceof AST_Return && !stat.value && ret.length == 0):
```

This seems to help me a lot :-)

avdg commented:
Got something working, but still have to write some tests. It involves checking if the statement above is an `AST_If` node containing `AST_Return` in body as well.

avdg commented:
Fix is up

kzc commented on lib/compress.js:
```diff
@@ -601,12 +601,18 @@ merge(Compressor.prototype, {
+                        //
+                        // Also require to at least have 2 return statements to counter the cost of the added `void `
                         if (compressor.option("sequences")
+                            && i > 0 && statements[i - 1] instanceof AST_If && statements[i - 1].body instanceof AST_Return
```
I don't think this change is sound or works in the general case. `i` should not be used here.

avdg replied:
the `i` is needed to check the statement above. Statements below are already processed and put in `ret`

avdg replied:
Also, this case is only called on the `if` statement on the bottom. This case is actually only called once per grouped if's.

kzc replied:
There could be intervening non-if statements that are not taken into account. This change looks like guess work to me.

avdg replied:
In case that there is no if statement above, there is only 1 return statement. This case will add a `void` in front of the tailing last statement. This cost must be compensated.

There might be a case that there is a `return` statement above, but this is dead code handling.

avdg replied:
Hmm... the only cases I can come up with are loops...

kzc replied:
The change also fails to take `else` clauses into account.

avdg replied:
Adding tests

kzc replied:
Unfortunately I don't have time to create any additional counter-examples.

kzc replied:
Initially I was concerned that `statements` was being altered elsewhere in the handle_if_return loop as it was being traversed backwards and `statements[i]` might be invalid. That doesn't appear to be the case. 

If another contributor is fine with this patch, by all means proceed.

kzc commented:
I think this issue is more complex than it first appears and requires more analysis. I would not rush to make a change based on a single counter-example. There are certainly other cases we have not considered.

avdg commented:
Build failed due to a single timeout :/

avdg commented:
Otherwise, I have no indications on real code that there is any downsides on the patch. Will do additional testing (I've modified the babili comparison tool to add my dev uglify, sadly it doesn't show bytes but maybe I can look into that a big deeper).

avdg commented:
Tests are only run once as speed is not the main issue

Testing with jquery.js on the babili tool.

Modifications to the test runner:

- Added uglify-dev, the setup is the patch mentioned at the top
- Added bytes after kb to make the smallest bytes visible

```
Patch as of 0331cb9
                    raw           raw win gzip          gzip win parse time run time (average)
uglify              85.15kB 87190 206%    29.73kB 30444 160%     8ms        1819ms
uglify-dev          85.14kB 87179 206%    29.73kB 30447 160%     9ms        1901ms
closure             85.7kB 87752  204%    30.47kB 31198 153%     9ms        2608ms
closure js          86.98kB 89065 200%    30.9kB 31645  150%     9ms        11507ms
babili (best speed) 93.75kB 95999 178%    32.08kB 32850 141%     9ms        5593ms
babili (best size)  93.75kB 95999 178%    32.08kB 32850 141%     9ms        5007ms

Removed optimization
                    raw           raw win gzip          gzip win parse time run time (average)
uglify              85.15kB 87190 206%    29.73kB 30444 160%     8ms        1795ms
uglify-dev          85.14kB 87179 206%    29.73kB 30447 160%     8ms        1737ms
closure             85.7kB 87752  204%    30.47kB 31198 153%     16ms       2658ms
closure js          86.98kB 89065 200%    30.9kB 31645  150%     9ms        10853ms
babili (best speed) 93.75kB 95999 178%    32.08kB 32850 141%     9ms        5422ms
babili (best size)  93.75kB 95999 178%    32.08kB 32850 141%     8ms        4879ms

No patch applied
                    raw           raw win gzip          gzip win parse time run time (average)
uglify              85.15kB 87190 206%    29.73kB 30444 160%     8ms        1770ms
uglify-dev          85.15kB 87190 206%    29.73kB 30444 160%     8ms        1724ms
closure             85.7kB 87752  204%    30.47kB 31198 153%     10ms       2518ms
closure js          86.98kB 89065 200%    30.9kB 31645  150%     8ms        12222ms
babili (best speed) 93.75kB 95999 178%    32.08kB 32850 141%     9ms        5539ms
babili (best size)  93.75kB 95999 178%    32.08kB 32850 141%     9ms        4846ms
```

avdg commented:
The worst one is the 3 bytes added on the gzip, while there is an improvement of 11 chars on the raw output.

kzc commented:
You have to take at least a dozen popular libraries into account for an accurate assessment. I find once the result is gzipped the patch is a wash on local files I've tried.

avdg commented:
uglify-dev refers to commit 0331cb9 in the cases below. (tests shamelessly taken over from https://github.com/babel/babili)

backbone
```
                    raw           raw win gzip        gzip win parse time run time (average)
uglify-dev          21.92kB 22449 222%    7.32kB 7496 170%     3ms        445ms
uglify              21.93kB 22461 222%    7.33kB 7510 170%     2ms        444ms
closure             21.83kB 22356 223%    7.4kB 7579  168%     2ms        1539ms
babili (best speed) 22.18kB 22708 218%    7.54kB 7722 163%     3ms        1242ms
babili (best size)  22.18kB 22708 218%    7.54kB 7722 163%     3ms        858ms
closure js          24.17kB 24745 192%    8.1kB 8298  144%     2ms        4685ms
```

react react/dist/react.js
```
                    raw           raw win gzip          gzip win parse time run time (average)
closure             34.77kB 35603 259%    11.96kB 12252 149%     4ms        1781ms
uglify              36.14kB 37006 245%    12.11kB 12401 146%     3ms        692ms
uglify-dev          36.14kB 37006 245%    12.11kB 12401 146%     3ms        660ms
babili (best speed) 35.67kB 36529 250%    12.49kB 12791 139%     3ms        1757ms
babili (best size)  35.67kB 36529 250%    12.49kB 12791 139%     3ms        1391ms
closure js          65.41kB 66980 91%     15.83kB 16215 88%      4ms        942ms
```

three three/build/three.js
```
                    raw             raw win gzip            gzip win parse time run time (average)
closure             491.56kB 503358 104%    124.93kB 127924 60%      35ms       6226ms
uglify-dev          497.75kB 509698 101%    125.08kB 128084 60%      35ms       5932ms
uglify              497.77kB 509718 101%    125.13kB 128135 60%      34ms       5488ms
closure js          499.79kB 511787 101%    126kB 129019    59%      35ms       77994ms
babili (best speed) 525.88kB 538497 91%     129.91kB 133025 54%      36ms       15509ms
babili (best size)  525.88kB 538497 91%     129.91kB 133025 54%      37ms       15442ms
```

avdg commented:
yeah, gzip kinda jumps randomly

kzc commented:
Can you please edit out the non-uglify rows as they are not of interest to this PR?

avdg commented:
Removed the kb as well as they are distracting.

kzc commented:
Regarding the test timeouts - could you please double the number in this mocha test:

https://github.com/mishoo/UglifyJS2/blob/e51c6ba38014fc73a4804a69c556d96f777bc0b3/test/mocha/cli.js#L7

kzc commented:
@avdg - If you and @rvanvelzen are okay with this patch, I'm good with it.

avdg commented:
I'll squash

kzc commented on test/mocha/cli.js:
```diff
@@ -55,7 +55,7 @@ describe("bin/uglifyjs", function () {
            if (err) throw err;
 
-           assert.strictEqual(stdout, "var bar=function(){function foo(bar){return bar}return foo}();\n" + 
+           assert.strictEqual(stdout, "var bar=function(){function foo(bar){return bar}return foo}();\n" +
```
Another PR in the queue changed this same line. Could you leave the trailing space in this one just to avoid the hassle of merging later?

avdg replied:
Sure, and yeah I've wondered why it changed. And yeah it was a space and luckily no other newline type.

avdg replied:
Actually it's a non-issue if both alters are the same, but will do as I don't have the other patch and can't be sure 100%.

avdg replied:
Thanks for notifying

kzc commented on test/compress/if_return.js:
```diff
@@ -170,8 +170,51 @@ if_return_7: {
+            else fail(e);
+        }
+
+        function i(e) {
```
How is the test `function i(e)` different from `function g(e)`?

avdg replied:
I think the else clause has to be removed

avdg replied:
That is done by now

kzc commented on test/compress/if_return.js:
```diff
@@ -170,8 +170,51 @@ if_return_7: {
+        if_return:    true,
+        sequences:    true,
+        conditionals: true,
+        side_effects : true,
```
alignment: remove space before `:`

avdg replied:
Done

kzc commented:
> Only add `void ` when a `return` keyword can be removed

That description doesn't match what the patch does.

Suggested commit name: "Better optimize if/return when no trailing return in block."

avdg commented:
What about current title?

avdg commented:
uh `in` is so confusing, so changing that into `after`

kzc commented:
Please keep the title simple: "Better optimize if/return when no trailing return in block."

Or even: "Better optimize if/return"

avdg commented:
Uh yeah, language is hard sometimes...

alexlamsl commented:
@avdg @kzc may I offer the following (on top of `master`) as an alternative?

It passes all the new & modified tests the same way, and may be easier to reason with.

```diff
diff --git a/lib/compress.js b/lib/compress.js
index bbd3659..7df30bb 100644
--- a/lib/compress.js
+++ b/lib/compress.js
@@ -602,20 +602,14 @@ merge(Compressor.prototype, {
                             continue loop;
                         }
                         //---
-                        // XXX: what was the intention of this case?
-                        // if sequences is not enabled, this can lead to an endless loop (issue #866).
-                        // however, with sequences on this helps producing slightly better output for
-                        // the example code.
-                        if (compressor.option("sequences")
-                            && ret.length == 1 && in_lambda && ret[0] instanceof AST_SimpleStatement
-                            && (!stat.alternative || stat.alternative instanceof AST_SimpleStatement)) {
+                        // if (a) return b; if (c) return d; e; ==> return a ? b : c ? d : void e;
+                        if (compressor.option("sequences") && in_lambda && ret.length === 2
+                            && ret[0] instanceof AST_If && ret[0].body instanceof AST_Return
+                            && ret[1] instanceof AST_SimpleStatement) {
                             CHANGED = true;
-                            ret.push(make_node(AST_Return, ret[0], {
-                                value: make_node(AST_Undefined, ret[0])
+                            ret.push(make_node(AST_Return, ret[1], {
+                                value: make_node(AST_Undefined, ret[1])
                             }).transform(compressor));
-                            ret = as_statement_array(stat.alternative).concat(ret);
-                            ret.unshift(stat);
-                            continue loop;
                         }
                     }
```

avdg commented:
Sounds promising, will take a deeper look when I have more time

avdg commented:
It does about exactly the same, but is slightly more understandable (no `stat` and `statements` confusing here).

I actually didn't knew that switch was that flexible in js btw. Especially with languages like c being more strict (and probably java too but my brain isn't wired to that case).

alexlamsl commented:
@avdg IIRC Java only supported primitive types (`boolean`, `char`, `short`, `int`, `long`, `float` & `double`), then expanded to `Enum` in 1.5, and eventually `String` in 7.

Even then, you can only specify constants in the `case` statements.

avdg commented:
Hmm I didn't read the code beyond the `if (stat.body instanceof AST_Return) {`.

So the part I don't understand is why the `continue loop` isn't needed. But the tests are passing so there must be some logic there to handle it well.

I just want to be sure that there isn't any nasty things before I remove the [wip] flag again.

avdg commented:
Well I do understand that there is a chain of transformation in this code, making things a bit hard to understand on its own.

It's more like that you need an engine and power for a car. But you only see the car and don't understand you need both engine and energy to make it go forward.

avdg commented:
uh sorry for the patch noise

kzc commented:
@alexlamsl @avdg regarding the suggested patch: https://github.com/mishoo/UglifyJS2/pull/1437#issuecomment-275328573 that was incorporated into this PR, why was the following code removed? 
```
-                            ret = as_statement_array(stat.alternative).concat(ret);
-                            ret.unshift(stat);
-                            continue loop;
```
In particular the removal of `continue loop;` changes the subsequent control flow considerably in ways I do not understand. This was not the end of a `case` statement where `continue loop;` would have been redundant.

`function handle_if_return` is the most complicated piece of code in Uglify. We must be very cautious here.

The previous patch from @avdg (despite my initial objections) deviated less from the field-tested logic.

I am **-1** on this PR as it stands.

avdg commented:
I would not understand that removing more code should block a pr. But I would understand that the logics/mechanics are too hard to understand.

This needs more documentation.

avdg commented:
Not to forget: the logics will have to be changed again on harmony once again to deal with block scoped variables.

alexlamsl commented:
@avdg @kzc let me see if I can explain the logic properly :sweat_smile: 

[`handle_if_returns()`](https://github.com/mishoo/UglifyJS2/blob/7f8d72d9d37396f2da05d5d824f74bd414c30119/lib/compress.js#L545) takes an array of `AST_Statement`s and outputs an (optimised) array of the same type, which can only be [`ret`](https://github.com/mishoo/UglifyJS2/blob/7f8d72d9d37396f2da05d5d824f74bd414c30119/lib/compress.js#L677).

The [for-loop](https://github.com/mishoo/UglifyJS2/blob/7f8d72d9d37396f2da05d5d824f74bd414c30119/lib/compress.js#L550) builds up `ret` by processing the input array backwards, which means at any point in the loop we are dealing with the current statement as `stat` and with `ret` containing the (optimised) tail of statements.

So within the [lines of change](https://github.com/mishoo/UglifyJS2/pull/1437/files#diff-230c53e13e3ee367aec7d49470eb42caR617) we are dealing with `stat` which is [guaranteed](https://github.com/mishoo/UglifyJS2/blob/7f8d72d9d37396f2da05d5d824f74bd414c30119/lib/compress.js#L559-L560) to be of the form `if (a) return b;` (or `if (a) return b; else <body>;`).

Unlike all the previous blocks of optimisations, however, this one does not modify `stat` at all - so I see it the same way as if `stat` misses all the if-return` rules and carry on to the rest of the for-loop body. Adding `ret.unshift(stat); continue loop;` works just fine, in case you prefer this as being less intrusive a change.

kzc commented:
> I would not understand that removing more code should block a pr.

You cannot remove code indiscriminately without understanding its effects. This early Uglify code does not have comprehensive code coverage in tests.

avdg commented:
>> I would not understand that removing more code should block a pr.

> You cannot remove code indiscriminately without understanding its effects. This early Uglify code does not have comprehensive code coverage in tests.

(uh excuse me, the context is that we still have valid tests and transformations are still very specified, but the chain of transformation is still unclear to me. As far as I'm concerned changing code and not being able to write a failing test and not understanding the code at the same time isn't a good indication of good code development)

alexlamsl commented:
Update: my most recent suggestion of adding back `continue loop;` failed one of the field tests I run into.

About to run through the same set of tests with the current version in this PR.

avdg commented:
> handle_if_returns() takes an array of AST_Statements and outputs an (optimised) array of the same type, which can only be ret.

> The for-loop builds up ret by processing the input array backwards, which means at any point in the loop we are dealing with the current statement as stat and with ret containing the (optimised) tail of statements.

This is what I do get so far

> So within the lines of change we are dealing with stat which is guaranteed to be of the form if (a) return b; (or if (a) return b; else <body>;).

> Unlike all the previous blocks of optimisations, however, this one does not modify stat at all - so I see it the same way as if stat misses all the if-return rules and carry on to the rest of the for-loop body. Adding ret.unshift(stat); continue loop;` works just fine, in case you prefer this as being less intrusive a change.

I guess the guarantees should be documented, these aren't obvious (at least to me) and there are many cases that have to conform to them.

kzc commented:
> Unlike all the previous blocks of optimisations, however, this one does not modify stat at all - so I see it the same way as if stat misses all the if-returnrules and carry on to the rest of the for-loop body.

But `ret` was modified. And the fall through code also modifies `ret`.

> Addingret.unshift(stat); continue loop;` works just fine, in case you prefer this as being less intrusive a change.

That would be preferable, although I was more comfortable with the previous code @avdg had.

I'm going to step back from reviewing Uglify code in general and you guys can sort it out. I urge caution changing this code in subtle ways that are not entirely understood. It may be better to leave it alone.

I'd ask @rvanvelzen to carefully review this code before merging it. I suggest the next release bump the minor number to 2.8 so people can fall back on using 2.7.x if need be.

alexlamsl commented:
@kzc I fully agree with your cautious approach. If anything I'm against the change as well if the scope has proven to be too difficult to comprehend.

@avdg given all these loops and inner functions, I guess we'll need to document & write tests for this part properly before making changes to them.

Just so we are on the same page (& correct me if I'm wrong) - this PR is aiming to solve one problem, i.e. preventing `if (a) return b; c;` getting turned into `return a?b:void c;`.

alexlamsl commented:
@avdg is your previous version under https://github.com/avdg/UglifyJS2/tree/if_return-backup ? I'll test it with the tests I've just got failing on my version of changes.

And let me apologise for barging in with broken code changes - I shall be more careful before making suggestions next time.

avdg commented:
> Just so we are on the same page (& correct me if I'm wrong) - this PR is aiming to solve one problem, i.e. preventing if (a) return b; c; getting turned into return a?b:void c;.

Yes, but preferably in an understandable way.

> @avdg is your previous version under https://github.com/avdg/UglifyJS2/tree/if_return-backup ? I'll test it with the tests I've just got failing on my version of changes.

Yes, but not squashed.

avdg commented:
> And let me apologise for barging in with broken code changes - I shall be more careful before making suggestions next time.

I do think it's fine to challenge other opinions. But yeah, doing so in a way that may result in anything bad (especially in production) should be avoided.

I think it's ok to fail on pr though, as this would prove the code acceptance is low when code is messy or unclear.

kzc commented:
@alexlamsl No worries. It's all part of the process.

My analysis could be wrong. You may understand the code better than I do. I was just expressing my concerns.

alexlamsl commented:
@avdg grrr... of course I confused `2.7.4` with `master` :man_facepalming: 

The current changeset as it is on this PR works fine on my tests - I guess two wrongs make it a right :rofl: 

(Obviously that doesn't mean this is bug free, but at least it is no longer definitely wrong...)

avdg commented:
I think I would add comments on the two latter cases (below the if block in the case).

Also motivating why there is no need (in this case) for following piece is needed:
```diff
-                            ret = as_statement_array(stat.alternative).concat(ret);
-                            ret.unshift(stat);
-                            continue loop;
```

But I will take time understanding the code before I write any comments down.

avdg commented:
Other than that controversial chance, the only difference between your patch and my patch is that yours is checking with an `if` statement in the `ret` list and mines does it before.

alexlamsl commented:
FYI, this changeset shaves 41 bytes off `angular.js` - I'll go through the bytes to ensure they are all sensible after shower...

@avdg I'll test adding back `ret.unshift(stat); continue loop;` in a moment as well, as I believe it wouldn't make a difference to the result (+ `npm test` still passes).

avdg commented:
I think the `ret.unshift(stat)` has something to do with the statement no? Since your patch is checking statements 1 off, I would not do that at all...

avdg commented:
Anyway, this is good proof of how little we understand this code anyway.

avdg commented:
From https://github.com/mishoo/UglifyJS2/pull/1437#issuecomment-275421716

> But ret was modified. And the fall through code also modifies ret.

Don't forget we are in a loop, And that many cases are modifying `ret`

alexlamsl commented:
@avdg so if every optimisation attempt failed, the [default operation](https://github.com/mishoo/UglifyJS2/blob/7f8d72d9d37396f2da05d5d824f74bd414c30119/lib/compress.js#L670-L674) is `ret.unshift(stat);`

In this particular block we have not modified `stat`, so we should this statement on verbatim to be returned by `handle_if_return()`.

avdg commented:
oh yeah I see, the `ret.unshift` comes in pairs with going to the next statement.

alexlamsl commented:
One thing that currently bugs me is in the original source, after [prepending `stat.alternative` to `ret`](https://github.com/mishoo/UglifyJS2/blob/0610c020b1544820be9898a285ab6c9066490552/lib/compress.js#L622), it just prepend `stat` to `ret` without clearing its `alternative`.

So it feels like `ret` reads `if (a) return b; else <alternative body>; <alternative body>; ...`

kzc commented:
>> But ret was modified. And the fall through code also modifies ret.

> Don't forget we are in a loop, And that many cases are modifying ret

I'm aware of that. I was referring to the fall through code immediately following the removed `continue loop` which further modifies `ret`.

> One thing that currently bugs me is in the original source, after prepending stat.alternative to ret, it just prepend stat to ret without clearing its alternative.

It did not have to clear the alternative due to the "continue loop".

@avdg I think you should revert to using your lower risk PR from yesterday that only added extra constraints to that particular transform rather than changing the logic of the entire function:

diff from master:
```
                         if (compressor.option("sequences")
+                            && i > 0 && statements[i - 1] instanceof AST_If && statements[i - 1].body instanceof AST_Return
                             && ret.length == 1 && in_lambda && ret[0] instanceof AST_SimpleStatement
                             && (!stat.alternative || stat.alternative instanceof AST_SimpleStatement)) {
                             CHANGED = true;
```

alexlamsl commented:
@kzc even with the `continue loop;` it still matters because `stat` and its `alternative` are now included in the `ret` array. Case in point, this fails on current `master`:
```js
issue_1431: {
    options = {
        if_return    : true,
        sequences    : true
    }
    input: {
        function f() {
            if (a())
                return b();
            else
                c();
            d();
        }
    }
    expect: {
        function f() {
            if (a())
                return b();
            else
                return c(), d(), void 0;
        }
    }
}
```
without adding `stat.alternative = null;` the current `master` will give:
```js
function f(){
    if(a())
        return b();
    else c();
    return c(),d(),void 0;
}
```

kzc commented:
@alexlamsl I stand corrected. You've found a genuine bug in the uglify-js@2.7.5 release:

$ cat t.js 
```
function f() {
    if (a())
        return b();
    else
        c();
    d();
}
```
$ uglifyjs t.js -c -b
```
function f() {
    return a() ? b() : (c(), void d());
}
```

alexlamsl commented:
@kzc actually I can't see the bug with your example... :sweat_smile: 

I'm currently going through v2.7.5 to see if I can narrow down with default compress option masked this bug.

kzc commented:
> @kzc actually I can't see the bug with your example... 

Blech, you're right. It's fine. I missed the return in `return b()`.

I give up.

alexlamsl commented:
@kzc ... but you did help me greatly with your example, because it's `conditionals` which is masking this bug (and which is on by default).

@avdg so taking [@kzc 's point](https://github.com/mishoo/UglifyJS2/pull/1437#issuecomment-275461858) of minimal change into account, shall we go with your original one-liner, plus `stat.alternative = null;` [before this line](https://github.com/mishoo/UglifyJS2/blob/7f8d72d9d37396f2da05d5d824f74bd414c30119/lib/compress.js#L623), alongside my test case above?

alexlamsl commented:
Meanwhile I shall hunt down any potential bugs in `AST_IF` and `compress.conditionals` :man_health_worker:

kzc commented:
uglify-js 2.7.5

$ cat t.js
```
function f() {
    if (a())
        return b();
    else
        c();
    d();
}
```
$ uglifyjs t.js -b -c conditionals=false
```
function f() {
    if (a()) return b(); else c();
    return c(), void d();
}
```
@alexlamsl I think we can agree that this is a bug in the 2.7.5 release?

avdg commented:
I don't actually know what I'm going to include now. There are many unknowns but every solution passes the tests. And also these changes involve lots of options as well.

I'm tending to do a 'fight for the fittest' now. But if that takes too much resources, I may prefer the simple fix then.

avdg commented:
Just a heads up, I might be busy this weekend. So things may get up delayed anyway.

kzc commented:
Because @alexlamsl predicted and demonstrated a genuine bug in the release (with non-default options), I endorse whatever recommendation he makes.

alexlamsl commented:
@kzc yup, you've got the bug right there :+1: 

@avdg take your time - I need to read up more around this to get more confident about the changes anyway

I have a feeling something in `conditionals` is also buggy, not that it necessarily can be exposed to end users. But I'll have a look as well - if nothing else I get to be more familiar with the codebase.

kzc commented:
The interaction of various non-default compress settings has never been tested. It's really a use-at-own-risk sort of situation. But then again, this software has no guarantees either. 

If the tests were to be run with random compress settings where the minified code is executed and compared to the expected result of a non-minified version it would lead to a greater measure of confidence in the software. Presently the minified test code is not executed, which I think is a shortcoming.

avdg commented:
I will include the test case though now I can. It can be removed anytime if things diverse too much though.

avdg commented:
@kzc unit testing tend to promote to test small pieces of code, but here it's not an option with the current code. Except playing with the options. This I think needs to be reviewed on it's own.

avdg commented:
And fuzzy code generation isn't a bad idea, but I don't think these needs to be in the default tests. Preferably as a separate tool. Unless we found bugging test cases ofc...

alexlamsl commented:
@avdg just to answer your question about what to include - my current thinking (against master, only code changes):
```diff
--- a/lib/compress.js
+++ b/lib/compress.js
@@ -613,6 +613,7 @@ merge(Compressor.prototype, {
                         // however, with sequences on this helps producing slightly better output for
                         // the example code.
                         if (compressor.option("sequences")
+                            && i > 0 && statements[i - 1] instanceof AST_If && statements[i - 1].body instanceof AST_Return
                             && ret.length == 1 && in_lambda && ret[0] instanceof AST_SimpleStatement
                             && (!stat.alternative || stat.alternative instanceof AST_SimpleStatement)) {
                             CHANGED = true;
@@ -620,6 +621,7 @@ merge(Compressor.prototype, {
                                 value: make_node(AST_Undefined, ret[0])
                             }).transform(compressor));
                             ret = as_statement_array(stat.alternative).concat(ret);
+                            stat.alternative = null;
                             ret.unshift(stat);
                             continue loop;
                         }
```
Then, in addition to your new tests in this PR, add this:
```diff
--- a/test/compress/if_return.js
+++ b/test/compress/if_return.js
@@ -205,3 +205,27 @@ issue_1089: {
         }
     }
 }
+
+issue_1431: {
+    options = {
+        if_return    : true,
+        sequences    : true
+    }
+    input: {
+        function f() {
+            if (a())
+                return b();
+            else
+                c();
+            d();
+        }
+    }
+    expect: {
+        function f() {
+            if (a())
+                return b();
+            else
+                return c(), d(), void 0;
+        }
+    }
+}
```

kzc commented:
The proposed execute tests would not be the same as existing tests. They'd be more akin to the test262 tests with asserts but minified with random uglify options.

alexlamsl commented:
@kzc I actually quite like your testing methodology of evaluating the minified result to see if they give the same answer. Obviously for `function()`s we might need to specify the range of input parameters to test as well.

That and a benchmark script to download and run against major JS libraries so as to ensure performance are on the back of my head as TODOs.

alexlamsl commented:
For post-mortem, the reason why `conditionals=true` masks the `if_return` bug is due to [this rule](https://github.com/mishoo/UglifyJS2/blob/0610c020b1544820be9898a285ab6c9066490552/lib/compress.js#L1953-L1961): `if (a) return b; else c;` :arrow_right: `if (a) return b; c;`

So as long as `conditionals=true`, `stat.alternative` [will always be](https://github.com/mishoo/UglifyJS2/blob/0610c020b1544820be9898a285ab6c9066490552/lib/compress.js#L617) `null`, thus never triggering the bug.

avdg commented:
> +issue_1431: {

I guess this must be issue_1437?

alexlamsl commented:
@avdg yup, sorry for the typo :sweat:

avdg commented:
No worries :-)

alexlamsl commented:
Now the question is, shall we simplify the condition to just `!stat.alternative` (thus removing `ret = as_statement_array(stat.alternative).concat(ret);` instead of adding `stat.alternative = null`) to fix this bug instead?

This will make this particular optimisation dependent on `conditionals` as well for the case of `if (a) return b; if (c) return d; else e; f;`, but given that it's on by default most users won't notice this.

avdg commented on lib/compress.js:
```diff
@@ -607,19 +607,23 @@ merge(Compressor.prototype, {
                             && ret.length == 1 && in_lambda && ret[0] instanceof AST_SimpleStatement
                             && (!stat.alternative || stat.alternative instanceof AST_SimpleStatement)) {
                             CHANGED = true;
                             ret.push(make_node(AST_Return, ret[0], {
```
I still had some dangling old code and accidently kept it in the code but this is quite interesting:

Changing `ret[0]` into `ret[1]` doesn't make the test fail.

alexlamsl replied:
That second parameter in `make_node()` on makes a non-essential reference to its original node, i.e. it doesn't affect the result of `AST_Node.print_to_string()` regardless of what you put there.

I have yet to trace down what uses that part in the codebase, but AFAICT it doesn't break any optimisations that I know of.

avdg replied:
uh, it's just token tracking, we don't have tests for that anyway...

avdg replied:
(and wow, I didn't saw your comment the whole time apparenty)

alexlamsl commented:
Just to illustrate what I meant above in code (on top of this PR):
```diff
--- a/lib/compress.js
+++ b/lib/compress.js
@@ -617,13 +617,11 @@ merge(Compressor.prototype, {
                         if (compressor.option("sequences")
                             && i > 0 && statements[i - 1] instanceof AST_If && statements[i - 1].body instanceof AST_Return
                             && ret.length == 1 && in_lambda && ret[0] instanceof AST_SimpleStatement
-                            && (!stat.alternative || stat.alternative instanceof AST_SimpleStatement)) {
+                            && !stat.alternative) {
                             CHANGED = true;
                             ret.push(make_node(AST_Return, ret[0], {
                                 value: make_node(AST_Undefined, ret[0])
                             }).transform(compressor));
-                            ret = as_statement_array(stat.alternative).concat(ret);
-                            stat.alternative = null;
                             ret.unshift(stat);
                             continue loop;
                         }
```
Adjust tests:
```diff
--- a/test/compress/if_return.js
+++ b/test/compress/if_return.js
@@ -272,7 +272,32 @@ issue_1437: {
             if (c())
                 return d();
             else
-                return e(), f(), void 0;
+                e();
+            f();
+        }
+    }
+}
+
+issue_1437_conditionals: {
+    options = {
+        conditionals : true,
+        if_return    : true,
+        sequences    : true
+    }
+    input: {
+        function x() {
+            if (a())
+                return b();
+            if (c())
+                return d();
+            else
+                e();
+            f();
+        }
+    }
+    expect: {
+        function x() {
+            return a() ? b() : c() ? d() : (e(), f(), void 0);
         }
     }
 }
```
So my rationale for this, apart from less code, is to remove duplicated functionality.

avdg commented:
> -                return e(), f(), void 0;
+                e();
+            f();

That wasn't good code anyway.

avdg commented:
Pushed new fix

alexlamsl commented:
I think this is ready for merge now, fixing a bug and improving an optimisation :sunglasses:

avdg commented:
uh I wished to add more comments, for the latter two cases.

alexlamsl commented:
@avdg go right ahead. I don't think we are in a hurry to push the fix since as @kzc it only affects non-default usage (plus nobody reported an issue on this yet).

Give me a shout if you need any further assistance.

kzc commented on test/compress/if_return.js:
```diff
@@ -205,3 +248,56 @@ issue_1089: {
+issue_1437: {
+    options = {
+        if_return    : true,
+        sequences    : true
```
Can you add the option `conditionals : false,` to this test just to emphasize where the bug existed in the 2.7.5 release?

avdg replied:
yikes, you hit the jackpot...

avdg replied:
nvm I just typed in the option wrongly. Resulting in a unreadable error.

avdg replied:
Done

avdg commented:
Uh, that timeout causing build failure seems to be bumped up in harmony only.

alexlamsl commented:
For me testing time-outs happened twice on PRs and [once on `master`](https://travis-ci.org/mishoo/UglifyJS2/builds/195495503) so far. 

Though only on Travis CI, as I never run into that problem locally.

avdg commented:
I'm going to remove the wip flag for now. Since there is a believe that fixing with including a edge test cases (and few more tests) shall suffice and I don't feel my part of ownership on this part of code, I'm going to leave it this way.

Maybe the easiest thing I can do is diffing between minified js code we found that have been impacted the most.
